### PR TITLE
Pluck Vertex

### PR DESCRIPTION
**Pull request by GeorgeNash**
Pluck a value from a tensor, this plucks the relevant auto diff too if it's a double.

I need this before I can do `if`. https://improbableio.atlassian.net/secure/RapidBoard.jspa?rapidView=377&modal=detail&selectedIssue=SOL-1976

I also tidied up the concat dual number method.

### CODE REVIEW

File: keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/DualNumber.java [hunk 1]
```diff
@@ -309,5 +320,34 @@ private DoubleTensor concatPartialDerivates(int dimension, List<DoubleTensor> pa
...
+
+        for (Map.Entry<Long, DoubleTensor> entry : this.partialDerivatives.asMap().entrySet()) {
+            DoubleTensor plucked = pluckedFromPartial(entry.getValue(), inputLength, flatIndex);
+            pluckedDuals.put(entry.getKey(), plucked);
+        }
+
+        return new DualNumber(DoubleTensor.scalar(this.value.getValue(index)), pluckedDuals);
+    }
+
+    private DoubleTensor pluckedFromPartial(DoubleTensor partial, long inputLength, int inputFlatIndex) {
```
[the remark this answers is not in the document]
Comment (GeorgeNash) [reply keanu-project/src/main/java/io/improbable/keanu/vertices/dbl/nonprobabilistic/diff/DualNumber.java]: Awesome man, I've introduced this

File: keanu-project/src/main/java/io/improbable/keanu/vertices/intgr/nonprobabilistic/operators/unary/IntegerPluckVertex.java [hunk 2]
```diff
@@ -0,0 +1,34 @@
...
+    private final int[] index;
+
+    /**
+     * A vertex that extracts a scalar at a given index
+     *
+     * @param inputVertex the input vertex to extract from
+     * @param index the index to extract at
+     */
+    public IntegerPluckVertex(IntegerVertex inputVertex, int... index) {
+        super(Tensor.SCALAR_SHAPE, inputVertex);
```
[the remark this answers is not in the document]
Comment (GeorgeNash) [reply keanu-project/src/main/java/io/improbable/keanu/vertices/intgr/nonprobabilistic/operators/unary/IntegerPluckVertex.java]: cool - I've added a validator

File: keanu-project/src/main/java/io/improbable/keanu/tensor/TensorShapeValidation.java [hunk 3]
```diff
@@ -111,4 +111,17 @@ public static void checkTensorsMatchNonScalarShapeOrAreScalar(int[] proposalShap
         return concatShape;
     }
 
+    public static void checkIndexIsValid(int[] shape, int... index) {
+        for (int i = 0; i < index.length; i++) {
```
Comment (gordoncaleb) [inline keanu-project/src/main/java/io/improbable/keanu/tensor/TensorShapeValidation.java]: Check the length of the shape and index as well?

File: keanu-project/src/test/java/io/improbable/keanu/tensor/generic/SimpleTensorTest.java [hunk 4]
```diff
@@ -12,6 +20,38 @@
...
+        DataBuffer buffer = Nd4j.createBuffer(data);
+        INDArray rank4 = Nd4j.create(buffer, shape);
+
+        INDArray plucked = pluck(rank4, 0, 1);
+        assertArrayEquals(new int[]{2, 2, 2}, plucked.shape());
+        assertArrayEquals(new double[]{5, 6, 7, 8, 9, 10, 11, 12}, plucked.data().asDouble(), 0.0);
+    }
+
+
+    private INDArray pluck(INDArray from, int... indices) {
```
Comment (gordoncaleb) [inline keanu-project/src/test/java/io/improbable/keanu/tensor/generic/SimpleTensorTest.java]: This pluck function exists in the DualNumber class already right?

Comment (GeorgeNash) [reply keanu-project/src/test/java/io/improbable/keanu/tensor/generic/SimpleTensorTest.java]: Oops, meant to delete!

**Pull request merged**